src/main/app.js

```diff
@@ -497,6 +505,28 @@ fluid.onUncaughtException.addListener(function (err) {
     }
 }, "gpii.app", "last");
 
+/**
+ * Handles the onMessage event of the gpii.windows.messages component.
+ *
+ * @param that {Component} An instance of gpii.app
+ * @param hwnd {number} Window handle.
+ * @param msg {number} The message.
+ * @param wParam {number} Message parameter.
+ * @param lParam {number} Message parameter.
+ * @param result {object} Set a 'value' field to specify a return value.
+ */
+gpii.app.windowMessage = function (that, hwnd, msg, wParam, lParam, result) {
+    // https://msdn.microsoft.com/library/aa376890
+    var WM_QUERYENDSESSION = 0x11;
+    if (msg === WM_QUERYENDSESSION) {
+        fluid.log(fluid.logLevel.FATAL, "System shutdown detected.");
+        that.exit();
```

Comment:
Amazingly economical... 

I guess we're protected by the general node.js behaviour that the process will not exit whilst there is some outstanding I/O? I'm wondering what happens in the case this request is outstanding for a while - https://github.com/stegru/gpii-app/blob/d99017a19b4b1a76dd6426b85621f4d74588206d/src/main/app.js#L399 

Reading this, it is actually the WM_ENDSESSION message which operates the shutdown, and so it's great to give ourselves a head start - https://blogs.msdn.microsoft.com/oldnewthing/20130627-00/?p=3973 - but it says "If you need to do cleanup operations before termination, you need to do them before returning from the WM_END­SESSION message". Which suggests that this cleanup needs to be synchronous, somehow - is there any way we could possibly block the message queue until we have actually done our cleanup?

Reply:
Yes, WM_ENDSESSION is the correct one to use, and no we can't block it.

Once the function returns, that's it - you've accepted death. Windows kills the process shortly after.

`that.exit()` does some async stuff, so you need to return in order for it to do it's stuff. I can't return a promise to win32. I tried blocking with Sleep (hoping something magical would happen), but that just pauses everything as you'd expect.

So, this responds to WM_QUERYENDSESSION with "please don't shutdown", which gives it a chance to die with grace. (WM_ENDSESSION doesn't accept a "don't do it" response). When GPII finally dies in its own time, Windows continues with the shutdown.

It does mean that if another application (like unsaved notepad) stalls the shutdown, and the use could cancel the shutdown, then they're left with no GPII - at least they'll still have high-contrast enabled ;) (now that I've thought of that, I'll make it so GPII gets the touch of death last)

Reply:
hmm.. I could "peek" the queue, and block it if the WM_ENDSESSION is found, before officially receiving the message.

Reply:
This looks like the mega-guide: https://msdn.microsoft.com/en-us/library/ms700677%28v=vs.85%29.aspx

Given you return 0 to WM_QUERYENDSESSION I guess this is what gets converted to "false" meaning that we want to block shutdown.
It seems that
i) We should in this case provide a reason via ShutdownBlockReasonCreate
ii) We should only block shutdown if someone is actually keyed in
It seems a real shame that we can't do anything async since it seems that the timeouts we are allowed (5 seconds and 30 seconds) are pretty generous and easily enough to do anything that we need....  weren't you experimenting in the old days with some weird kind of "thread pump" stuff when you were thinking of working with named pipes?

Reply:
Yes, I tried ShutdownBlockReasonCreate - that prevented receiving the two *ENDSESSION messages (feels odd - I need to double check). I don't want to prevent shutdown, just delay it.

I already have a message pump (for running in gpii-windows) https://github.com/stegru/windows/blob/GPII-2869/gpii/node_modules/windowMessages/src/windowMessages.js#L157

My plan is to use that.

Reply:
No, there is already a massage loop (provided by chromium) - it will just become a race.
I will make GPII receive WM_QUERYENDSESSION last (after every other window has handled it), meaning at that point every other application has agreed to shutdown, and if it weren't for GPII then WM_ENDSESSION would be sent out.